File: scripts/trial/actions/count_until_server.py
#!/usr/bin/env python
import rospy
import actionlib

from agv_motion.msg import CountUntilAction
from agv_motion.msg import CountUntilGoal
from agv_motion.msg import CountUntilResult
from agv_motion.msg import CountUntilFeedback

class CountUntilServer:
    def __init__(self):  # constructor
        # initalize simple action server
        self._as = actionlib.SimpleActionServer(
            "/count_until", 
            CountUntilAction, 
            execute_cb=self.on_goal, # callback on goal
            auto_start=False)  # so that it is started manually
        self._as.start()
        self._counter = 0
        rospy.loginfo("Simple Action Server has been started")

    # with this function class can receive goals
    def on_goal(self, goal):
        rospy.loginfo("A goal has been received")
        rospy.loginfo(goal)

        max_number = goal.max_number
        wait_duration = goal.wait_duration

        self._counter = 0
        rate = rospy.Rate(1.0/wait_duration)

        success = False
        preempted = False

        while not rospy.is_shutdown(): # self._counter < max_number:
            self._counter += 1

            if self._as.is_preempt_requested(): # handle preempt request
                preempted = True
                break

            if self._counter > 9: # handle goal abortion
                break

            if self._counter >= max_number: # handle goal success
                success = True
                break

            rospy.loginfo(self._counter)
            
            feedback = CountUntilFeedback()
            feedback.percentage = round((float(self._counter) / float(max_number)) * 100.0, 1)
            self._as.publish_feedback(feedback) # send feedback to the client

            rate.sleep()

        result = CountUntilResult()
        result.count = self._counter  # server finished handling the goal
        rospy.loginfo("Send goal result to client")

        if preempted:
            rospy.loginfo("Preempted")
            self._as.set_preempted(result)
        elif success:
            rospy.loginfo("Success")
            # result will be sent when the goal state is set to succeeded
            self._as.set_succeeded(result)  # result is sent back using topic
        else:
            rospy.loginfo("Failure")
            self._as.set_aborted(result)



if __name__ == '__main__':
    rospy.init_node("count_until_server")
    server = CountUntilServer()
    rospy.spin()

File: src/trial/actions/count_until_server.cpp
#include <ros/ros.h>
#include <actionlib/server/simple_action_server.h>

#include <agv_motion/CountUntilAction.h>
#include <agv_motion/CountUntilGoal.h>
#include <agv_motion/CountUntilResult.h>
#include <agv_motion/CountUntilFeedback.h>

class CountUntilServer
{

protected:
    ros::NodeHandle _nh;
    actionlib::SimpleActionServer<agv_motion::CountUntilAction> _as;
    int _counter;

public:
    CountUntilServer() : _as(_nh,
                             "/count_util",
                             boost::bind(&CountUntilServer::onGoal, this, _1),
                             false),
                         _counter(0)
    {
        _as.start();
        ROS_INFO("Simple Action Server has been started");
    }

    void onGoal(const agv_motion::CountUntilGoalConstPtr &goal)
    {
        ROS_INFO("Goal received");

        int max_number = goal->max_number;
        double wait_duration = goal->wait_duration;

        ROS_INFO("max_number: %d, wait_duration: %lf", max_number, wait_duration);

        _counter = 0;

        ros::Rate rate(1.0/wait_duration); // Hz

        bool success = false;
        bool preempted = false;

        while (ros::ok()) {
            
            _counter++;

            if (_as.isPreemptRequested()) {
                preempted = true;
                break;
            }

            if (_counter > 9) {
                break;
            }

            if (_counter > max_number) {
                success = true;
                break;
            }

            ROS_INFO("%d", _counter);

            // publish feedback
            agv_motion::CountUntilFeedback feedback;
            feedback.percentage = (double)_counter / (double)max_number;
            _as.publishFeedback(feedback);

            rate.sleep();
        } // while

        agv_motion::CountUntilResult result;
        result.count = _counter;
        ROS_INFO("Send goal result to client");

        if (preempted) {
            ROS_INFO("Preempted");
            _as.setPreempted(result);
        }
        else if (success) {
            ROS_INFO("Success");
            _as.setSucceeded(result);
        }
        else {
            ROS_INFO("Aborted");
            _as.setAborted(result);
        }
    }
};

int main(int argc, char **argv)
{
    ros::init(argc, argv, "count_until_server");

    ROS_INFO("Starting server...");

    CountUntilServer server;

    ros::spin();
}